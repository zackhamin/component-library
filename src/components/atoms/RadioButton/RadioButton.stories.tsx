import { IRadioButtonProps, RadioButton } from './RadioButton';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'atoms/RadioButton',
	component: RadioButton,
	argTypes: { onChange: { action: 'changed' } },
} as Meta;

const Template: Story<IRadioButtonProps> = (args) => {
	return <RadioButton {...args}>Child</RadioButton>;
};

export const defaultStory = Template.bind({});
defaultStory.args = {
	children: 'Test',
	name: 'name Test',
	value: 'Value 1',
	isDisabled: false,
	id: 'idTest',
	isChecked: false,
};

export const disabledOption = Template.bind({});
disabledOption.args = {
	children: 'Test Disabled',
	name: 'Disabled Test',
	value: 'Value 1',
	isDisabled: true,
	id: 'idTest',
	isChecked: false,
};

export const requiredOption = Template.bind({});
requiredOption.args = {
	children: 'Test Required',
	name: 'Required Test',
	value: 'Value 1',
	isDisabled: false,
	id: 'idTest',
	isChecked: false,
	isRequired: true,
};
