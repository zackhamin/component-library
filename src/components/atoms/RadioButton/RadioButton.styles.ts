import styled from 'styled-components';
import { colors, typography } from 'Styles';

const StyledItemDiv = styled.div`
	display: flex;
	align-items: center;
	position: relative;
	margin-bottom: 2rem;
`;

const StyledRadioButtonLabel = styled.label`
	font-family: ${typography.type.body};
	font-weight: 500;
	font-size: ${typography.size.s3 + 'rem'};
	padding: 1rem;
	margin-right: 1.3rem;
`;

const StyledRadioButton = styled.input`
	background-color: white;
	cursor: pointer;
	appearance: unset;
	border: 1px solid ${colors.mediumGrey};
	border-radius: 50%;
	width: 2.2rem;
	height: 2.2rem;
	&:checked {
		border: 1px solid ${colors.emerald};
		background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'%3e%3ccircle r='2.8' fill='%2300594C'/%3e%3c/svg%3e");
	}

	& + ${StyledRadioButtonLabel} {
		position: relative;
		display: inline-block;
		cursor: pointer;
	}
	${(props) =>
		props.disabled &&
		` 
    &:disabled {
		cursor: default;
			background: ${colors.disabled};
			cursor: default;

      }
  `}
`;

export { StyledItemDiv, StyledRadioButton, StyledRadioButtonLabel };
