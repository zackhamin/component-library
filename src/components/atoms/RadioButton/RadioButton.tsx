import { RadioButtonGroupContext } from 'Molecules/RadioButtonGroup/RadioButtonGroup';
import { ReactInputProps } from '@Types/form';
import React, { forwardRef, useContext } from 'react';
import {
	StyledItemDiv,
	StyledRadioButton,
	StyledRadioButtonLabel,
} from './RadioButton.styles';

export interface IRadioButtonProps extends Omit<ReactInputProps, 'ref'> {
	children: React.ReactNode;
	name?: string;
	value: string;
	onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
	isDisabled?: boolean;
	id: string;
	isChecked?: boolean;
	isRequired?: boolean;
}

const RadioButton = forwardRef<HTMLInputElement, IRadioButtonProps>(
	(
		{
			children,
			name,
			value,
			isDisabled,
			id,
			isChecked,
			onChange,
			isRequired,
			...rest
		},
		ref
	) => {
		const ctx = useContext(RadioButtonGroupContext);
		const radioCtx = {
			name: ctx?.name || name,
			onChange: ctx?.handleOnChange || onChange,
			isSelected: isChecked || ctx?.isSelected === value,
			isRequired: ctx?.isRequired || isRequired,
		};
		return (
			<StyledItemDiv>
				<StyledRadioButton
					id={id}
					type="radio"
					name={radioCtx.name}
					value={value}
					onChange={radioCtx.onChange}
					disabled={isDisabled}
					{...(!ctx && !isChecked ? {} : { checked: radioCtx.isSelected })}
					required={isRequired}
					ref={ref}
					{...rest}
				/>
				<StyledRadioButtonLabel htmlFor={id}>{children}</StyledRadioButtonLabel>
			</StyledItemDiv>
		);
	}
);

export { RadioButton };
