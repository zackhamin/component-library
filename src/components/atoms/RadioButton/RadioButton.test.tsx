import { RadioButton } from './RadioButton';
import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps = {
	name: 'Radio1',
	value: 'opt1',
	isDisabled: false,
	id: 'opt1',
	onChange: jest.fn(),
	children: 'label1',
};

describe('RadioButton', () => {
	it('renders  correctly', () => {
		const { container } = render(<RadioButton {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});
	it('renders with disabled state', () => {
		const withDisabled = { ...defaultProps, isDisabled: true };
		render(<RadioButton {...withDisabled} />);
		expect(screen.getByRole('radio')).toBeDisabled();
	});
});

describe('button click handlers', () => {
	it('fires change events on change of value', () => {
		render(<RadioButton {...defaultProps} />);
		const input = screen.getByDisplayValue('opt1');
		fireEvent.click(input, { target: { value: 'opt1' } });
		expect({ ...defaultProps.onChange }).toHaveBeenCalled();
	});
});
