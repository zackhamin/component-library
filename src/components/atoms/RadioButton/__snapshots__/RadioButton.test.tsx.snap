// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton renders  correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  margin-bottom: 2rem;
}

.c3 {
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 1.6rem;
  padding: 1rem;
  margin-right: 1.3rem;
}

.c1 {
  background-color: white;
  cursor: pointer;
  -webkit-appearance: unset;
  -moz-appearance: unset;
  appearance: unset;
  border: 1px solid #bebebe;
  border-radius: 50%;
  width: 2.2rem;
  height: 2.2rem;
}

.c1:checked {
  border: 1px solid #00594C;
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'%3e%3ccircle r='2.8' fill='%2300594C'/%3e%3c/svg%3e");
}

.c1 + .c2 {
  position: relative;
  display: inline-block;
  cursor: pointer;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      id="opt1"
      name="Radio1"
      type="radio"
      value="opt1"
    />
    <label
      class="c2 c3"
      for="opt1"
    >
      label1
    </label>
  </div>
</div>
`;
